### DISCUSSION TITLE
Deepspeed seems completely broken on Ubuntu 22.04.

### DISCUSSION THREAD

**Discussion by AlpsAficionado** [oobabooga/text-generation-webui#2022] [General]
Also posted this issue to the Deepspeed project itself at https://github.com/microsoft/DeepSpeed/discussions/3531, so much of this is a copypasta:

I've spent several hours bashing my head against getting deepspeed to function properly on my system. I run oobabooga/text-generation-webui inside an Ubuntu VM on my server (Nvidia Quadro RTX 8000 with 48 VRAM; 128GB system RAM; the VM is given total control of the 8000 via PCI passthrough/iommu, and has 96GB of system RAM allocated to it).

After hours of frustration, I was unable to get deepspeed to function with oobabooga on Ubuntu 22.04.

I was able to get it to run on Ubuntu 23.04; however, 22.04 is the current long-term supported version (and it still took me significant manual intervention to get it going in 23.04).

A rundown of my issues with 22.04:

1) Error: `ModuleNotFoundError: No module named 'deepspeed'`
Solution: `pip install deepspeed`

2) Error: ModuleNotFoundError: No module named 'mpi4py'
Solution: `sudo apt install libopenmpi-dev ; pip install mpi4py`

3) Error: 'pip install mpi4py' won't work; it crashes like so:

```
/home/user/oobabooga_linux/installer_files/env/compiler_compat/ld: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so: undefined reference to \opal_class_finalize'
/home/user/oobabooga_linux/installer_files/env/compiler_compat/ld: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so: undefined reference to \mca_base_var_group_get_count'
/home/user/oobabooga_linux/installer_files/env/compiler_compat/ld: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so: undefined reference to \opal_datatype_resize'
/home/user/oobabooga_linux/installer_files/env/compiler_compat/ld: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so: undefined reference to \opal_hash_table_set_value_uint64'
/home/user/oobabooga_linux/installer_files/env/compiler_compat/ld: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so: undefined reference to \MPIR_being_debugged'
/home/user/oobabooga_linux/installer_files/env/compiler_compat/ld: /usr/lib/x86_64-linux-gnu/openmpi/lib/libmpi.so: undefined reference to \opal_list_t_class'
collect2: error: ld returned 1 exit status
failure.
removing: _configtest.c _configtest.o
error: Cannot link MPI programs. Check your configuration!!!
```

Solution: `env LD_LIBRARY_PATH=/usr/lib/x86_64-linux-gnu/ pip install --no-cache-dir mpi4py`

4) Error:
```
module = importlib.util.module_from_spec(spec)
File "<frozen importlib._bootstrap>", line 571, in module_from_spec
File "<frozen importlib._bootstrap_external>", line 1176, in create_module
File "<frozen importlib._bootstrap>", line 241, in _call_with_frames_removed
ImportError: /home/user/oobabooga_linux/installer_files/env/bin/../lib/libstdc++.so.6: version \GLIBCXX_3.4.30' not found (required by /home/user/.cache/torch_extensions/py310_cu117/async_io/async_io.so)
```

Solution: conda install -c conda-forge gcc=12.1.0 (rebuilds/reinstalls a whole bunch of crap, see below:

```
Downloading and Extracting Packages
_libgcc_mutex-0.1 | 3 KB | ###################################################################################### | 100%
ca-certificates-2023 | 145 KB | ###################################################################################### | 100%
gcc_impl_linux-64-12 | 56.2 MB | ####################################################################################2 | 98%
libgomp-12.2.0 | 455 KB | ###################################################################################### | 100%
kernel-headers_linux | 707 KB | ###################################################################################### | 100%
libgcc-ng-12.2.0 | 931 KB | ###################################################################################### | 100%
libgcc-devel_linux-6 | 3.3 MB | ###################################################################################### | 100%
openssl-1.1.1t | 1.9 MB | ###################################################################################### | 100%
requests-2.29.0 | 55 KB | ###################################################################################### | 100%
libsanitizer-12.1.0 | 6.2 MB | ###################################################################################### | 100%
libstdcxx-ng-12.2.0 | 4.3 MB | ###################################################################################### | 100%
sysroot_linux-64-2.1 | 31.4 MB | ####################################################################################6 | 98%
gcc-12.1.0 | 24 KB | ###################################################################################### | 100%
certifi-2023.5.7 | 149 KB | ###################################################################################### | 100%
_openmp_mutex-4.5 | 23 KB | ###################################################################################### | 100%
binutils_impl_linux- | 5.2 MB | ###################################################################################### | 100%
```

5) Error: RuntimeError: Unable to JIT load the async_io op due to it not being compatible due to hardware/software issue.

Full error spew:

```
INFO:Loading decapoda-research_llama-65b-hf...
Traceback (most recent call last):
File "/home/user/oobabooga_linux/text-generation-webui/server.py", line 919, in <module>
shared.model, shared.tokenizer = load_model(shared.model_name)
File "/home/user/oobabooga_linux/text-generation-webui/modules/models.py", line 117, in load_model
model = LoaderClass.from_pretrained(Path(f"{shared.args.model_dir}/{model_name}"), torch_dtype=torch.bfloat16 if shared.args.bf16 else torch.float16)
File "/home/user/oobabooga_linux/installer_files/env/lib/python3.10/site-packages/transformers/models/auto/auto_factory.py", line 471, in from_pretrained
return model_class.from_pretrained(
File "/home/user/oobabooga_linux/installer_files/env/lib/python3.10/site-packages/transformers/modeling_utils.py", line 2624, in from_pretrained
init_contexts = [deepspeed.zero.Init(config_dict_or_path=deepspeed_config())] + init_contexts
File "/home/user/oobabooga_linux/installer_files/env/lib/python3.10/site-packages/deepspeed/runtime/zero/partition_parameters.py", line 762, in __init__
self.param_swapper = AsyncPartitionedParameterSwapper(_ds_config, self.dtype)
File "/home/user/oobabooga_linux/installer_files/env/lib/python3.10/site-packages/deepspeed/runtime/swap_tensor/partitioned_param_swapper.py", line 40, in __init__
aio_op = AsyncIOBuilder().load(verbose=False)
File "/home/user/oobabooga_linux/installer_files/env/lib/python3.10/site-packages/deepspeed/ops/op_builder/builder.py", line 445, in load
return self.jit_load(verbose)
File "/home/user/oobabooga_linux/installer_files/env/lib/python3.10/site-packages/deepspeed/ops/op_builder/builder.py", line 449, in jit_load
raise RuntimeError(
RuntimeError: Unable to JIT load the async_io op due to it not being compatible due to hardware/software issue. None
```

Solution: I don't know; this is where I am stuck. https://github.com/microsoft/DeepSpeed/issues/1037 suggests that I just need to 'apt install libaio-dev', but I've done that and it doesn't help.

This is just a typical Linux rabbithole of playing whac-a-mole with dependency hell, precisely the sort of thing the one-click installer is supposed to obviate the need for.
I'm still stuck and cannot for the life of me get the --deepspeed option working.
I'd truly appreciate any help, and if anyone is smart enough to integrate a reliable solution to this issue into the one-click installer script, I would be quite grateful.
Thanks to all in advance!

**Comment by mironkraft** [comment]
My first thought will be to install ubuntu in another SSD directly, not using VM but, you can try next things:

1. **Use a Virtual Environment**: If you are not already doing so, I recommend setting up a virtual environment for your project to avoid conflicts between different versions of packages installed on your system. You can create a new Python virtual environment using `virtualenv` or `conda`. This way, all dependencies will be installed locally in the virtual environment rather than system-wide, reducing the chance of conflicts.

2. **Check DeepSpeed Installation**: The issue might be coming from the DeepSpeed installation. It's required to be installed from source when using it with PyTorch. Try reinstalling DeepSpeed from source following the instructions in the [official DeepSpeed installation guide](https://www.deepspeed.ai/docs/installation/#installing-from-source).

3. **Check PyTorch and CUDA Compatibility**: Ensure that the installed PyTorch and CUDA versions are compatible. DeepSpeed uses CUDA extensions, and they must be compiled with the same CUDA version as PyTorch. If they're not, you might face some issues.

4. **Rebuild DeepSpeed CUDA extensions**: The error you're encountering seems to be related to DeepSpeed's custom CUDA extensions. After you have ensured that CUDA and PyTorch versions are compatible, you might need to rebuild the DeepSpeed CUDA extensions. Follow these steps:

   - Remove the `~/.cache/torch_extensions` directory to clear the cached compiled extensions.
   - Reinstall DeepSpeed.

5. **Update GCC**: The issue might be related to the version of GCC installed in your system. Updating GCC might solve the issue. Try using the latest version of GCC that's compatible with your CUDA version.

6. **Use Docker**: If all else fails, you might consider using Docker. DeepSpeed provides a [Dockerfile](https://www.deepspeed.ai/docs/installation/#docker-container) that can be used to create a container with all the necessary dependencies installed.

Remember to always activate your virtual environment before running any Python commands, to ensure that you're using the correct versions of your packages. But this are the "easy things", check what i thing you shoudl go for this:

It appears you have a complex problem related to setting up DeepSpeed in your Ubuntu VM environment. I will try to help you as much as I can.

Here are a few potential solutions for the issues you have encountered.

### ImportError: /home/user/oobabooga_linux/installer_files/env/bin/../lib/libstdc++.so.6: version \GLIBCXX_3.4.30' not found

The issue is coming from the fact that your `libstdc++.so.6` does not have the `GLIBCXX_3.4.30` version, which is required by the package you are trying to use.

You can check the versions of `GLIBCXX` that are available in your `libstdc++.so.6` by running the following command:

```bash
strings /home/user/oobabooga_linux/installer_files/env/bin/../lib/libstdc++.so.6 | grep GLIBCXX
```

If you can't find `GLIBCXX_3.4.30` in the output, you will have to update your GCC. You did attempt to install a newer GCC version using Conda, but it seems to have not solved the problem.

One possible issue might be that the new GCC's libraries are not being used. You can check which `libstdc++.so.6` is being used with `ldconfig -p | grep stdc++`. If the path is not the one from the new GCC, you will need to add it to the `LD_LIBRARY_PATH`.

```bash
export LD_LIBRARY_PATH=/path/to/new/gcc/libs:$LD_LIBRARY_PATH
```

Replace `/path/to/new/gcc/libs` with the path to the new GCC's libraries.

### RuntimeError: Unable to JIT load the async_io op due to it not being compatible due to hardware/software issue

This issue may be due to the environment not being correctly configured for JIT (Just-In-Time) compilation. The error indicates that DeepSpeed's `async_io` operation cannot be JIT compiled. This operation is used for asynchronous I/O operations and is likely necessary for DeepSpeed to function correctly.

One common cause of this issue is that the CUDA version installed in your environment is not compatible with the PyTorch version. You need to ensure that the CUDA toolkit version matches the CUDA version that your PyTorch was built with. You can check the CUDA version of PyTorch with `torch.version.cuda` and the system CUDA version with `nvcc --version`.

Another potential cause could be a version mismatch between your NVIDIA driver and the CUDA version. You can check your NVIDIA driver version with `nvidia-smi`.

If none of these solutions work, the issue could be specific to the DeepSpeed package or your hardware configuration. In this case, I recommend reaching out to the DeepSpeed community for further help, as you've already done.

But as I think, mostly if you go with that much power, best is to install not use VM from my point of view.

**Comment by poisenbery** [comment]
I'm actually having an issue with Deepspeed running on PopOS 22.04
Except my entire system hangs when I try to load a model, so I can't really do much to troubleshoot.
I used the 1 click installer and the instructions on the deepspeed section.

**Comment by wanghao-cst** [comment]
I got the same issue on Ubuntu not VM. Have no idea what to do next.
<img width="1370" alt="截屏2023-07-23 13 45 48" src="https://github.com/oobabooga/text-generation-webui/assets/55015183/3afb9dd7-67ff-44ea-b8c7-3aceafa7bd08">

**Comment by ghost** [comment]
sudo apt install libaio-dev
